Migrate CI to GitHub Actions

coalkettler opened the pull request:
Introduces a `continuous_integration` workflow to build and publish all Docker image variants to Quay.io. Replaces Travis CI configuration.

Resolves #112 

---

**Testing**

- Verify that [CI workflow](https://github.com/azavea/docker-django/pull/114/checks?sha=c7205db6e643d7e4d812ad9be8eb4bd63f694cfb) passed for all variants
- Verify that [this CI workflow run](https://github.com/azavea/docker-django/pull/114/checks?sha=7c8db979da1485dc2a0d67a46a8a5abc8b67847a) that was configured to run `cipublish` successfully published images to Quay
- Verify that Travis checks did not run for latest commit

coalkettler committed:
Create continuous integration workflow for building containers with GitHub Actions

coalkettler committed:
Publish images if running on master branch

coalkettler committed:
Test running cipublish for PR #114

coalkettler committed:
Revert "Test running cipublish for PR #114"

This reverts commit 7c8db979da1485dc2a0d67a46a8a5abc8b67847a.

coalkettler committed:
Remove Travis config

coalkettler committed:
Update example PostgreSQL major versions in README

coalkettler commented on .github/workflows/continuous_integration.yml:
```diff
@@ -0,0 +1,42 @@
+        django_version: ["2.2"]
+        python_version: ["3.6", "3.7", "3.8"]
+        variant: [alpine, slim]
+        include:
+          - variant: alpine
+            pg_major: 12
+            pg_version: 12.2-r0
+          - variant: slim
+            pg_major: 10
+            pg_version: 10.13-1.pgdg100+1
```
This is a [pretty cool feature](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-including-additional-values-into-combinations). Handy for setting the dependent variables, like which Postgres version we install, based on independent variables, like the variant.

coalkettler commented on .github/workflows/continuous_integration.yml:
```diff
@@ -0,0 +1,42 @@
+    strategy:
+      matrix:
+        django_version: ["2.2"]
+        python_version: ["3.6", "3.7", "3.8"]
```
Because Django doesn't maintain the same Python version compatibility between all versions, this isn't guaranteed to always work as part of the matrix. However it'll certainly work until we add the next LTS version, and if we're restricting ourselves to a common set of more recent Python versions supported by all LTS versions.

I think we should cross that bridge when we come to it and rely on something like `exclude` for edge cases, instead of getting really verbose with specifying Python versions per version in `include`.

If we wanted something closer to how Travis' matrix syntax works, we could be verbose and state everything as an object as [shown here](https://github.com/actions/runner/issues/343). However that feature is sparsely documented and I'd rather rely on the well-documented `property: [list]` syntax.

rbreslow replied:
> If we wanted something closer to how Travis' matrix syntax works, we could be verbose and state everything as an object as [shown here](https://github.com/actions/runner/issues/343). However that feature is sparsely documented and I'd rather rely on the well-documented `property: [list]` syntax.

As it is now, it's hard to understand what the matrix strategy is. Does it just test every possible permutation of combinations of each array?

coalkettler replied:
Yep! So it'd build Django `2.2` for Python `3.6`, `3.7`, and `3.8`, and then build each of those for `alpine` and `slim`, for a total of 6 builds.

Additionally, the `include` below will match on anything with `variant == alpine` and set the `PG_MAJOR` and `PG_VERSION` appropriately for each of the matching permutations. Same deal with `slim`.

coalkettler commented on .github/workflows/continuous_integration.yml:
```diff
@@ -0,0 +1,42 @@
+      - run: ./scripts/cipublish
+        if: github.ref == 'refs/heads/master'
+        env:
+          QUAY_USER: ${{ secrets.QUAY_IMAGES_USER }}
+          QUAY_PASSWORD: ${{ secrets.QUAY_IMAGES_PASSWORD }}
```
I have the credentials for our `azavea+images` user set as an organizational secret. I've granted access to those secrets to this repo and `azavea/docker-tippecanoe` (which we can update to use this secret if we end up using it here). I namespaced it with `IMAGES` since we have a few different machine users that we use for different application groups.

I think we should use organizational secrets in situations like this to make it easier to revoke / roll credentials that are used in multiple repositories.

coalkettler commented on README.md:
```diff
@@ -7,7 +7,7 @@ This repository contains a collection of templated `Dockerfile` for image varian
 
 - `PYTHON_VERSION` - Python interpreter version (for `python:*-slim-buster` or `python:*-alpine` [images](https://hub.docker.com/_/python/))
-- `PG_MAJOR` - Major version number of the target PostgreSQL database instance (`9.3`, `9.4`, `9.5`, `9.6`, and `10`)
+- `PG_MAJOR` - Major version number of the target PostgreSQL database instance (`10` and `12`)
```
Lil fixup here 🔧

coalkettler commented on .github/workflows/continuous_integration.yml:
```diff
@@ -0,0 +1,42 @@
+        variant: [alpine, slim]
+        include:
+          - variant: alpine
+            pg_major: 12
```
Using numbers instead of strings here because I think recent versions are always integers. Using strings for anything semver-ish.

coalkettler commented on .travis.yml:
```diff
@@ -1,23 +0,0 @@
-  - VERSION=2.2 PYTHON_VERSION=3.8 PG_MAJOR=12 PG_VERSION=12.2-r0 VARIANT=alpine
-
-before_script:
-  - wget -qO- 'https://github.com/tianon/pgp-happy-eyeballs/raw/master/hack-my-builds.sh' | bash
```
I noticed that we didn't include this in `azavea/docker-tippecanoe`. I gather that it's to make accessing PGP keyservers a little more consistent. Is this no longer beneficial?

rbreslow commented on .github/workflows/continuous_integration.yml:
```diff
@@ -0,0 +1,42 @@
+            pg_version: 10.13-1.pgdg100+1
+    env:
+      CI: 1
+      DOCKER_BUILDKIT: 1
```
What is this for?

coalkettler replied:
This opts into using the newer [Docker BuildKit enhancements](https://docs.docker.com/develop/develop-images/build_enhancements/) for Docker builds. They purport some new features and performance improvements. We've been using it in [docker-tippecanoe as well](https://github.com/azavea/docker-tippecanoe/blob/master/.github/workflows/continuous-integration.yml#L18).

rbreslow commented on .github/workflows/continuous_integration.yml:
```diff
@@ -0,0 +1,42 @@
+            pg_major: 10
+            pg_version: 10.13-1.pgdg100+1
+    env:
+      CI: 1
```
`CI` is always set to `true`.

See: https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables

coalkettler replied:
Oh fabulous, thank you! I'll submit a small PR to update this in [docker-tippecanoe](https://github.com/azavea/docker-tippecanoe/blob/master/.github/workflows/continuous-integration.yml#L19) as well.

rbreslow reviewed (commented):
Nice work on this. I'm sorry that it took me so long to finish my initial pass.

coalkettler committed:
Remove unnecessary CI variable

This variable is set to `true` by default: https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables

rbreslow reviewed (approved):
Nice job, Cole 👍
